### PR TITLE
AzureServiceBus ensure client reuse

### PR DESCRIPTION
**Pull request by TomMalow**
**What this PR does / why we need it**:
HealthChecks.AzureServiceBus did not reuse created clients clients when possible.
Using multiple health check to queues and subscription within the same ServiceBus would resolve in multiple clients being created. Thereby experience a long first time execution of the health check and unnecessary open connections.

**Which issue(s) this PR fixes**:
#1890 

**Special notes for your reviewer**:
I moved the code over to use the shared ClientCache instead of an internal cache. The internal cache did not allow for caching across health check instances.
The HealthChecks.AzureServiceBus did not have any unit test before. I have added for most of the health check except EventBus health check as it was not modified by this PR. This is due to the EventBus health check not face the same problem.

**Does this PR introduce a user-facing change?**: 
Yes:
- New constructors are added to all ServiceBus Health Checks to allow for injecting customServiceBus client . Mainly for testing purpose.
- A previous `Prefix` field as been made obsolete and superseded by the existing `ConnectionKey`

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [x] Code compiles correctly
- [x] Created/updated tests
- [x] Unit tests passing
- [x] End-to-end tests passing
- [ ] Extended the documentation - No prior documentation
- [ ] Provided sample for the feature - No prior samples

### CODE REVIEW

**Comment by TomMalow** [comment]
@dotnet-policy-service agree company="Delegate A/S"

File: src/HealthChecks.AzureServiceBus/DependencyInjection/AzureServiceBusHealthCheckBuilderExtensions.cs [hunk 1]
```diff
@@ -805,7 +805,7 @@ public static class AzureServiceBusHealthCheckBuilderExtensions
                 };
 
                 configure?.Invoke(options);
-                return new AzureServiceBusSubscriptionHealthCheck(options);
+                return new AzureServiceBusSubscriptionHealthCheck(options, new ServiceBusClientProvider());
```
Comment (sungam3r) [inline src/HealthChecks.AzureServiceBus/DependencyInjection/AzureServiceBusHealthCheckBuilderExtensions.cs:808]: For all calls with `new ServiceBusClientProvider()` you may use something like default static instance.

File: src/HealthChecks.AzureServiceBus/AzureServiceBusHealthCheck.cs [hunk 2]
```diff
@@ -7,21 +6,18 @@ namespace HealthChecks.AzureServiceBus;
...
-
-    protected static readonly ConcurrentDictionary<string, ServiceBusReceiver> ServiceBusReceivers = new();
-
     protected TOptions Options { get; }
 
+    private readonly ServiceBusClientProvider _clientProvider;
+
     protected string Prefix => Options.ConnectionString ?? Options.FullyQualifiedNamespace!;
 
     protected abstract string ConnectionKey { get; }
```
Comment (sungam3r) [inline src/HealthChecks.AzureServiceBus/AzureServiceBusHealthCheck.cs:15]: All implementation return `Prefix` from `ConnectionKey`. Let's make ConnectionKey virtual:
` protected virtual string ConnectionKey => Options.ConnectionString ?? Options.FullyQualifiedNamespace!;` and mark `Prefix` obsolete.

File: src/HealthChecks.AzureServiceBus/DependencyInjection/AzureServiceBusHealthCheckBuilderExtensions.cs (hunk 1 above)
Comment (sungam3r) [reply src/HealthChecks.AzureServiceBus/DependencyInjection/AzureServiceBusHealthCheckBuilderExtensions.cs:808]: Or you may return ctor with 1 argument for backward compatibility. Then no changes required in this file.

File: src/HealthChecks.AzureServiceBus/AzureServiceBusHealthCheck.cs (hunk 2 above)
Comment (TomMalow) [reply src/HealthChecks.AzureServiceBus/AzureServiceBusHealthCheck.cs:15]: On it. Was also not a fan of keeping `Prefix` to make as little change as possible as the name does not make sense.

File: src/HealthChecks.AzureServiceBus/DependencyInjection/AzureServiceBusHealthCheckBuilderExtensions.cs (hunk 1 above)
Comment (TomMalow) [reply src/HealthChecks.AzureServiceBus/DependencyInjection/AzureServiceBusHealthCheckBuilderExtensions.cs:808]: On it. For some reason I never excecuted the Api Approal Test. My bad.

File: src/HealthChecks.AzureServiceBus/AzureServiceBusHealthCheck.cs [hunk 3]
```diff
@@ -7,21 +6,19 @@ namespace HealthChecks.AzureServiceBus;
...
-
-    protected static readonly ConcurrentDictionary<string, ServiceBusAdministrationClient> ManagementClientConnections = new();
-
-    protected static readonly ConcurrentDictionary<string, ServiceBusReceiver> ServiceBusReceivers = new();
-
     protected TOptions Options { get; }
 
+    private readonly ServiceBusClientProvider _clientProvider;
+
+    [Obsolete]
```
Comment (sungam3r) [inline src/HealthChecks.AzureServiceBus/AzureServiceBusHealthCheck.cs:13]:
```suggestion
    [Obsolete("Please, override ConnectionKey property instead.")]
```

File: test/HealthChecks.AzureServiceBus.Tests/HealthChecks.AzureServiceBus.approved.txt [hunk 4]
```diff
@@ -8,40 +8,46 @@ namespace HealthChecks.AzureServiceBus
...
     {
-        protected static readonly System.Collections.Concurrent.ConcurrentDictionary<string, Azure.Messaging.ServiceBus.ServiceBusClient> ClientConnections;
-        protected static readonly System.Collections.Concurrent.ConcurrentDictionary<string, Azure.Messaging.ServiceBus.Administration.ServiceBusAdministrationClient> ManagementClientConnections;
-        protected static readonly System.Collections.Concurrent.ConcurrentDictionary<string, Azure.Messaging.ServiceBus.ServiceBusReceiver> ServiceBusReceivers;
-        protected AzureServiceBusHealthCheck(TOptions options) { }
-        protected abstract string ConnectionKey { get; }
+        protected AzureServiceBusHealthCheck(TOptions options, HealthChecks.AzureServiceBus.ServiceBusClientProvider clientProvider) { }
+        protected virtual string ConnectionKey { get; }
         protected TOptions Options { get; }
+        [System.Obsolete]
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/HealthChecks.AzureServiceBus.approved.txt:14]:
```suggestion
        [System.Obsolete("Please, override ConnectionKey property instead.")]
```

File: src/HealthChecks.AzureServiceBus/AzureServiceBusQueueHealthCheck.cs [hunk 5]
```diff
@@ -32,19 +34,20 @@ public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context
...
-        Task CheckWithReceiver()
+        async Task CheckWithReceiver()
         {
-            var client = ClientConnections.GetOrAdd(ConnectionKey, _ => CreateClient());
-            var receiver = ServiceBusReceivers.GetOrAdd(
-                $"{nameof(AzureServiceBusQueueHealthCheck)}_{ConnectionKey}",
-                client.CreateReceiver(Options.QueueName));
+            var client = await ClientCache.GetOrAddAsyncDisposableAsync(ConnectionKey, _ => CreateClient()).ConfigureAwait(false);
+            var receiver = await ClientCache.GetOrAddAsyncDisposableAsync(
+                $"{nameof(AzureServiceBusQueueHealthCheck)}_{queueKey}",
```
Comment (sungam3r) [inline src/HealthChecks.AzureServiceBus/AzureServiceBusQueueHealthCheck.cs:41]: Thit key is computed every time. Please move `{nameof(AzureServiceBusQueueHealthCheck)}` into `queueKey` and make it field.

File: src/HealthChecks.AzureServiceBus/AzureServiceBusSubscriptionHealthCheck.cs [hunk 6]
```diff
@@ -34,19 +35,20 @@ public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context
...
-        Task CheckWithReceiver()
+        async Task CheckWithReceiver()
         {
-            var client = ClientConnections.GetOrAdd(ConnectionKey, _ => CreateClient());
-            var receiver = ServiceBusReceivers.GetOrAdd(
-                $"{nameof(AzureServiceBusSubscriptionHealthCheck)}_{ConnectionKey}",
-                client.CreateReceiver(Options.TopicName, Options.SubscriptionName));
+            var client = await ClientCache.GetOrAddAsyncDisposableAsync(ConnectionKey, _ => CreateClient()).ConfigureAwait(false);
+            var receiver = await ClientCache.GetOrAddAsyncDisposableAsync(
+                $"{nameof(AzureServiceBusSubscriptionHealthCheck)}_{subscriptionKey}",
```
Comment (sungam3r) [inline src/HealthChecks.AzureServiceBus/AzureServiceBusSubscriptionHealthCheck.cs:42]: The same here.

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+            .Received(1)
+            .CreateManagementClient(ConnectionString);
+    }
+
+    [Fact]
+    public async Task reuses_existing_client_when_using_same_connection_string_with_different_topic()
+    {
+        // First call
+        using var tokenSource = new CancellationTokenSource();
+        var options = new AzureServiceBusQueueMessagesCountThresholdHealthCheckOptions(TopicName)
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs:63]: topic or queue?

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        int messageCount,
+        int degradedThreshold,
+        int unhealthyThreshold,
+        HealthStatus expectedHealthStatus)
+    {
+        using var tokenSource = new CancellationTokenSource();
+        var options = new AzureServiceBusQueueMessagesCountThresholdHealthCheckOptions(TopicName)
+        {
+            ConnectionString = ConnectionString,
+            ActiveMessages = new AzureServiceBusQueueMessagesCountThreshold()
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs:213]:
```suggestion
            ActiveMessages = new AzureServiceBusQueueMessagesCountThreshold
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+        int messageCount,
+        int degradedThreshold,
+        int unhealthyThreshold,
+        HealthStatus expectedHealthStatus)
+    {
+        using var tokenSource = new CancellationTokenSource();
+        var options = new AzureServiceBusQueueMessagesCountThresholdHealthCheckOptions(TopicName)
+        {
+            ConnectionString = ConnectionString,
+            ActiveMessages = new AzureServiceBusQueueMessagesCountThreshold(),
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs:260]:
```suggestion
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+            DeadLetterMessages = new AzureServiceBusQueueMessagesCountThreshold()
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs:261]:
```suggestion
            DeadLetterMessages = new AzureServiceBusQueueMessagesCountThreshold
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+                DegradedThreshold = degradedThreshold,
+                UnhealthyThreshold = unhealthyThreshold,
+            }
+        };
+        var healthCheck = new AzureServiceBusQueueMessageCountThresholdHealthCheck(options, _clientProvider);
+        var context = new HealthCheckContext
+        {
+            Registration = new HealthCheckRegistration(HealthCheckName, healthCheck, HealthStatus.Unhealthy, null)
+        };
+        var queueProperties = ServiceBusModelFactory.QueueRuntimeProperties(TopicName, activeMessageCount: 0, deadLetterMessageCount: messageCount);
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs:272]:
```suggestion
        var queueProperties = ServiceBusModelFactory.QueueRuntimeProperties(TopicName, deadLetterMessageCount: messageCount);
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs (hunk 7 above)
Comment (sungam3r) [reply test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs:63]: I suppose copy-pasted from `azureservicebussubscriptionhealthcheck_should` all over the file.

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+            Registration = new HealthCheckRegistration(HealthCheckName, healthCheck, HealthStatus.Unhealthy, null)
+        };
+
+        var actual = await healthCheck
+            .CheckHealthAsync(context, tokenSource.Token)
+            .ConfigureAwait(false);
+
+        // Second call
+        var otherTopicName = Guid.NewGuid().ToString();
+        var otherOptions = new AzureServiceBusQueueHealthCheckOptions(otherTopicName)
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs:103]: oops, wrong type and unused

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+        {
+            ConnectionString = ConnectionString,
+        };
+        var otherHealthCheck = new AzureServiceBusSubscriptionHealthCheck(options, _clientProvider);
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs:107]:
```suggestion
        var otherHealthCheck = new AzureServiceBusSubscriptionHealthCheck(otherOptions, _clientProvider);
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+        var otherContext = new HealthCheckContext
+        {
+            Registration = new HealthCheckRegistration(HealthCheckName, healthCheck, HealthStatus.Unhealthy, null)
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs:110]:
```suggestion
            Registration = new HealthCheckRegistration(HealthCheckName, otherHealthCheck, HealthStatus.Unhealthy, null)
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs (hunk 7 above)
Comment (TomMalow) [reply test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs:63]: Yeah, rush job from my side. Sorry about that.

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+        var otherTopicName = Guid.NewGuid().ToString();
+        var otherOptions = new AzureServiceBusQueueMessagesCountThresholdHealthCheckOptions(otherTopicName)
+        {
+            FullyQualifiedNamespace = FullyQualifiedName,
+            Credential = _tokenCredential,
+        };
+        var otherHealthCheck = new AzureServiceBusQueueMessageCountThresholdHealthCheck(options, _clientProvider);
+        var otherContext = new HealthCheckContext
+        {
+            Registration = new HealthCheckRegistration(HealthCheckName, healthCheck, HealthStatus.Unhealthy, null)
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs:152]:
```suggestion
            Registration = new HealthCheckRegistration(HealthCheckName, otherHealthCheck, HealthStatus.Unhealthy, null)
```

Please check all tests for such typos.

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+        };
+
+        var otherActual = await healthCheck
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueMessageCountThresholdHealthCheckTests.cs:155]:
```suggestion
        var otherActual = await otherHealthCheck
```

And this.

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs [hunk 17]
```diff
@@ -0,0 +1,404 @@
...
+        };
+
+        var otherActual = await healthCheck
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs:113]:
```suggestion
        var otherActual = await otherHealthCheck
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs [hunk 18]
```diff
@@ -0,0 +1,404 @@
...
+            .ConfigureAwait(false);
+
+        // Second call
+        var otherTopicName = Guid.NewGuid().ToString();
+        var otherOptions = new AzureServiceBusSubscriptionHealthCheckHealthCheckOptions(otherTopicName, SubscriptionName)
+        {
+            FullyQualifiedNamespace = FullyQualifiedName,
+            Credential = _tokenCredential,
+        };
+        var otherHealthCheck = new AzureServiceBusSubscriptionHealthCheck(options, _clientProvider);
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs:197]:
```suggestion
        var otherHealthCheck = new AzureServiceBusSubscriptionHealthCheck(otherOptions, _clientProvider);
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs [hunk 19]
```diff
@@ -0,0 +1,404 @@
...
+        var otherContext = new HealthCheckContext
+        {
+            Registration = new HealthCheckRegistration(HealthCheckName, healthCheck, HealthStatus.Unhealthy, null)
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs:200]:
```suggestion
            Registration = new HealthCheckRegistration(HealthCheckName, otherHealthCheck, HealthStatus.Unhealthy, null)
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs [hunk 20]
```diff
@@ -0,0 +1,404 @@
...
+        };
+
+        var otherActual = await healthCheck
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs:203]:
```suggestion
        var otherActual = await otherHealthCheck
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs [hunk 21]
```diff
@@ -0,0 +1,404 @@
...
+        _serviceBusClient
+            .Received(1)
+            .CreateReceiver(TopicName, SubscriptionName);
+
+        await _serviceBusReceiver
+            .Received(1)
+            .PeekMessageAsync(cancellationToken: tokenSource.Token)
+            .ConfigureAwait(false);
+
+        actual.Status.ShouldBe(HealthStatus.Unhealthy);
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusSubscriptionHealthCheckTests.cs:316]: I would move status check right after ` var actual = await healthCheck
            .CheckHealthAsync(context, tokenSource.Token)
            .ConfigureAwait(false);`

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs [hunk 22]
```diff
@@ -0,0 +1,253 @@
...
+using NSubstitute.ExceptionExtensions;
+
+namespace HealthChecks.AzureServiceBus.Tests;
+
+public class azureservicebustopichealthcheck_should
+{
+    private readonly string ConnectionString;
+    private readonly string FullyQualifiedName;
+    private readonly string TopicName;
+    private readonly string HealthCheckName = "unit-test-check";
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs:14]:
```suggestion
    private const string HEALTH_CHECK_NAME = "unit-test-check";
```

here and in other files

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs [hunk 23]
```diff
@@ -0,0 +1,253 @@
...
+            .CheckHealthAsync(context, tokenSource.Token)
+            .ConfigureAwait(false);
+
+        _clientProvider
+            .Received(1)
+            .CreateManagementClient(ConnectionString);
+    }
+
+    [Fact]
+    public async Task reuses_existing_client_when_using_same_connection_string_with_different_queue()
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs:58]:
```suggestion
    public async Task reuses_existing_client_when_using_same_connection_string_with_different_topic()
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs [hunk 24]
```diff
@@ -0,0 +1,253 @@
...
+            Registration = new HealthCheckRegistration(HealthCheckName, healthCheck, HealthStatus.Unhealthy, null)
+        };
+
+        var actual = await healthCheck
+            .CheckHealthAsync(context, tokenSource.Token)
+            .ConfigureAwait(false);
+
+        // Second call
+        var otherTopicName = Guid.NewGuid().ToString();
+        var otherOptions = new AzureServiceBusQueueHealthCheckOptions(otherTopicName)
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs:78]: again wrong type and unused variable

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs [hunk 25]
```diff
@@ -0,0 +1,253 @@
...
+        {
+            ConnectionString = ConnectionString,
+        };
+        var otherHealthCheck = new AzureServiceBusTopicHealthCheck(options, _clientProvider);
+        var otherContext = new HealthCheckContext
+        {
+            Registration = new HealthCheckRegistration(HealthCheckName, healthCheck, HealthStatus.Unhealthy, null)
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs:85]:
```suggestion
            Registration = new HealthCheckRegistration(HealthCheckName, otherHealthCheck, HealthStatus.Unhealthy, null)
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs [hunk 26]
```diff
@@ -0,0 +1,253 @@
...
+        };
+
+        var otherActual = await healthCheck
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs:88]:
```suggestion
        var otherActual = await otherHealthCheck
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs [hunk 27]
```diff
@@ -0,0 +1,253 @@
...
+            .ConfigureAwait(false);
+
+        // Second call
+        var otherTopicName = Guid.NewGuid().ToString();
+        var otherOptions = new AzureServiceBusTopicHealthCheckOptions(otherTopicName)
+        {
+            FullyQualifiedNamespace = FullyQualifiedName,
+            Credential = _tokenCredential,
+        };
+        var otherHealthCheck = new AzureServiceBusTopicHealthCheck(options, _clientProvider);
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs:148]:
```suggestion
        var otherHealthCheck = new AzureServiceBusTopicHealthCheck(otherOptions, _clientProvider);
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs [hunk 28]
```diff
@@ -0,0 +1,253 @@
...
+        var otherContext = new HealthCheckContext
+        {
+            Registration = new HealthCheckRegistration(HealthCheckName, healthCheck, HealthStatus.Unhealthy, null)
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs:151]:
```suggestion
            Registration = new HealthCheckRegistration(HealthCheckName, otherHealthCheck, HealthStatus.Unhealthy, null)
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs [hunk 29]
```diff
@@ -0,0 +1,253 @@
...
+        };
+
+        var otherActual = await healthCheck
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusTopicHealthCheckTests.cs:154]:
```suggestion
        var otherActual = await otherHealthCheck
```

**Comment by sungam3r** [comment]
@TomMalow I'm fine to merge after fixing all copy-paste issues. Thank you very much for exhaustive tests.

**Comment by TomMalow** [comment]
@sungam3r It should be ready now. Found some other similar mistakes that have been correct.
Thanks for being so thorough!

**Comment by sungam3r** [comment]
Please bump ` <HealthCheckAzureServiceBus>7.0.0</HealthCheckAzureServiceBus>`.

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueHealthCheckTests.cs [hunk 30]
```diff
@@ -0,0 +1,402 @@
...
+            .CheckHealthAsync(context, tokenSource.Token)
+            .ConfigureAwait(false);
+
+        // Second call
+        var otherTopicName = Guid.NewGuid().ToString();
+        var otherOptions = new AzureServiceBusQueueHealthCheckOptions(otherTopicName)
+        {
+            ConnectionString = ConnectionString,
+        };
+        var otherHealthCheck = new AzureServiceBusQueueHealthCheck(options, _clientProvider);
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueHealthCheckTests.cs:105]:
```suggestion
        var otherHealthCheck = new AzureServiceBusQueueHealthCheck(otherOptions, _clientProvider);
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueHealthCheckTests.cs [hunk 31]
```diff
@@ -0,0 +1,402 @@
...
+        // Second call
+        var otherTopicName = Guid.NewGuid().ToString();
+        var otherOptions = new AzureServiceBusQueueHealthCheckOptions(otherTopicName)
+        {
+            ConnectionString = ConnectionString,
+        };
+        var otherHealthCheck = new AzureServiceBusQueueHealthCheck(otherOptions, _clientProvider);
+        var otherContext = new HealthCheckContext
+        {
+            Registration = new HealthCheckRegistration(HEALTH_CHECK_NAME, healthCheck, HealthStatus.Unhealthy, null)
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueHealthCheckTests.cs:108]:
```suggestion
            Registration = new HealthCheckRegistration(HEALTH_CHECK_NAME, otherHealthCheck, HealthStatus.Unhealthy, null)
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueHealthCheckTests.cs [hunk 32]
```diff
@@ -0,0 +1,402 @@
...
+        {
+            ConnectionString = ConnectionString,
+        };
+        var otherHealthCheck = new AzureServiceBusQueueHealthCheck(otherOptions, _clientProvider);
+        var otherContext = new HealthCheckContext
+        {
+            Registration = new HealthCheckRegistration(HEALTH_CHECK_NAME, otherHealthCheck, HealthStatus.Unhealthy, null)
+        };
+
+        var otherActual = await healthCheck
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueHealthCheckTests.cs:111]:
```suggestion
        var otherActual = await otherHealthCheck
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueHealthCheckTests.cs [hunk 33]
```diff
@@ -0,0 +1,402 @@
...
+            .ConfigureAwait(false);
+
+        // Second call
+        var otherQueueName = Guid.NewGuid().ToString();
+        var otherOptions = new AzureServiceBusQueueHealthCheckOptions(otherQueueName)
+        {
+            FullyQualifiedNamespace = FullyQualifiedName,
+            Credential = _tokenCredential,
+        };
+        var otherHealthCheck = new AzureServiceBusQueueHealthCheck(options, _clientProvider);
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueHealthCheckTests.cs:195]:
```suggestion
        var otherHealthCheck = new AzureServiceBusQueueHealthCheck(otherOptions, _clientProvider);
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueHealthCheckTests.cs [hunk 34]
```diff
@@ -0,0 +1,402 @@
...
+        var otherContext = new HealthCheckContext
+        {
+            Registration = new HealthCheckRegistration(HEALTH_CHECK_NAME, healthCheck, HealthStatus.Unhealthy, null)
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueHealthCheckTests.cs:198]:
```suggestion
            Registration = new HealthCheckRegistration(HEALTH_CHECK_NAME, otherHealthCheck, HealthStatus.Unhealthy, null)
```

File: test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueHealthCheckTests.cs [hunk 35]
```diff
@@ -0,0 +1,402 @@
...
+        };
+
+        var otherActual = await healthCheck
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/AzureServiceBusQueueHealthCheckTests.cs:201]:
```suggestion
        var otherActual = await otherHealthCheck
```

Please, look through test files one more time...

**Comment by TomMalow** [comment]
Got tired of embarrassing myself, so ended up refactoring the tests to reduce the potential for me making, or overseeing, any mistakes. The test now has reduced boilerplate around setting up and executing the HealthChecks. This also resulted in the tests being much easier to read.

**Comment by sungam3r** [comment]
@TomMalow I believe everything is OK now though I did not deep in each line this time.

File: test/HealthChecks.AzureServiceBus.Tests/HealthChecks.AzureServiceBus.approved.txt [hunk 36]
```diff
@@ -8,40 +8,46 @@ namespace HealthChecks.AzureServiceBus
     public abstract class AzureServiceBusHealthCheck<TOptions>
         where TOptions : HealthChecks.AzureServiceBus.Configuration.AzureServiceBusHealthCheckOptions
     {
-        protected static readonly System.Collections.Concurrent.ConcurrentDictionary<string, Azure.Messaging.ServiceBus.ServiceBusClient> ClientConnections;
```
Comment (sungam3r) [inline test/HealthChecks.AzureServiceBus.Tests/HealthChecks.AzureServiceBus.approved.txt:11]: Techically breaking changes, but I think v7.1 is OK.

Comment (sungam3r) [reply test/HealthChecks.AzureServiceBus.Tests/HealthChecks.AzureServiceBus.approved.txt:11]: b07157f

**Pull request merged**